File: CPP/leetcode_0747.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int dominantIndex(vector<int>& nums) {
        int size = nums.size();
        if (size <= 0) {
            return -1;
        }
        if (size == 1) {
            return 0;
        }
        int i = 0;
        int max = 0;
        int second = 0;
        int i_max = 0;
        for (int i = 0; i < size; i++) {
            if (i < size) {
                if (nums[i] > max) {
                    second = max;
                    i_max = i;
                    max = nums[i];
                }
                else if (nums[i] > second) {
                    second = nums[i];
                }
            }
        }
        if (max >= second * 2) {
            return i_max;
        }
        else {
            return -1;
        }  
    }
};

File: LeetCode/leetcode_0747.py
class Solution(object):
    def dominantIndex(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_index = -1
        second_max_value = -1
        for i in range(len(nums)):
            if i == 0:
                max_index = i
                continue
            value = nums[i]
            if value > nums[max_index]:
                value = nums[max_index]
                max_index = i
            if value > second_max_value:
                second_max_value = value
        if nums[max_index] >= (second_max_value * 2):
            return max_index
        else:
            return -1


if __name__ == '__main__':
    test_data1 = [3, 6, 1, 0]
    test_data2 = [1, 2, 3, 4]
    test_data3 = [2, 1]
    test_data4 = [1, 2]
    s = Solution()
    print(s.dominantIndex(test_data1))
    print(s.dominantIndex(test_data2))
    print(s.dominantIndex(test_data3))
    print(s.dominantIndex(test_data4))
